rhconsulting_schedules.rake

```diff
@@ -0,0 +1,188 @@
+    depot_attrs = export_file_depot(model_object_attrs).try(:attributes)
+    return model_object_attrs unless depot_attrs
+
+    parsed_depot_attrs = {
+      @file_depot_key => RhconsultingModelAttributes.parse_attributes(depot_attrs, :support_case)
+    }
+    model_object_attrs.merge!(parsed_depot_attrs)
+  end
+
+  # add zone attributes to the export so that we know how to search for it during import
+  # e.g. the zone_id does us no good because they are likely to be different
+  def add_zone_attrs(model_object_attrs)
+    zone       = export_zone(model_object_attrs)
+    zone_name  = zone ? zone.name : 'default'
+    zone_attrs = {
+      @zone_key => {
+        :name => zone_name
+      }
+    }
+
+    model_object_attrs.merge!(zone_attrs)
+  end
+
+  # normalize the following so that it is ready to be imported:
+  #   symbolize_keys: provide a consistent hash with symbols as keys
+  #   file_depot:     needed for import when DatabaseBackup type
+  #   zone:           name needed for import
+  #   options:        remove the options hash as all of the necessary data livees in the body
+  #                   and this is very inconsistent across schedule types (e.g. db vs. widget)
+  def normalize_export_data(raw_attrs)
+    raw_attrs_hash = raw_attrs.deep_symbolize_keys
+    add_file_depot_attrs(raw_attrs_hash) if raw_attrs_hash[:towhat] == 'DatabaseBackup'
+    add_zone_attrs(raw_attrs_hash)
+    raw_attrs_hash[:sched_action].delete(:options)
+    RhconsultingModelAttributes.parse_attributes(raw_attrs_hash, @rejected_attrs)
+  end
+
+  def export_schedules(export_dir, options)
+    schedules = if @widget_schedules
+                  MiqSchedule.all
+                else
+                  MiqSchedule.where('towhat != ?', 'MiqWidget')
+                end
+
+    schedules.each do |schedule|
+      # normalize the attributes from the model
+      normalized_attrs = normalize_export_data(schedule.attributes)
+
+      # set the filename and replace spaces and characters that are not allowed in filenames
+      fname = MiqIllegalChars.replace("#{schedule.name}.yaml", options)
+      File.write("#{export_dir}/#{fname}", normalized_attrs.to_yaml)
+    end
+  end
+
+  def normalize_import_data(raw_attrs)
+    user = export_user(raw_attrs)
+    zone = export_zone(raw_attrs)
+
+    # find and create the file depot if needed
+    if raw_attrs[:towhat] == 'DatabaseBackup'
+      file_depot   = export_file_depot(raw_attrs)
+      file_depot ||= FileDepot.create(raw_attrs[@file_depot_key])
+      raw_attrs.merge!(:file_depot_id => file_depot.id)
+    end
+
+    # update the options
+    raw_attrs.merge!(:userid => user.userid, :zone_id => zone.id)
+
+    # delete the additional data that our model does not need
+    [@file_depot_key, @zone_key].each { |key| raw_attrs.delete(key) }
+
+    # return the newly modified attrs
+    return raw_attrs
+  end
+
+  def import_schedules(import_dir)
+    MiqSchedule.transaction do
```

Comment:
I'd like to see the `#transaction` moved into the loop. That way if there is only one failed import in the directory the other schedules will be imported.